{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5f44886",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import pickle\n",
    "import sys\n",
    "sys.path.append(\"../../../Benchmark/\")\n",
    "from benchmarker import Benchmarker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ff91c33",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir(\"../../../datatset/MultiEmbryo2/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7530418f",
   "metadata": {},
   "source": [
    "Prepare data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9cb85f2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rna = sc.read_h5ad(\"rna_annot.h5ad\")\n",
    "# rna.obs_names = [i.split(\"-\")[0] for i in rna.obs_names]\n",
    "# rna.obs_names\n",
    "# # sc.pp.filter_genes(rna, min_counts=3)\n",
    "# rna.write(\"rna_annot2.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c856a567",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1949 × 32285\n",
       "    obs: 'annot'\n",
       "    var: 'gene_ids', 'feature_types', 'genome'\n",
       "    uns: 'annot_colors'\n",
       "    obsm: 'spatial'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna = sc.read_h5ad(\"rna_annot2.h5ad\")\n",
    "rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7b34cef2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm1992</th>\n",
       "      <td>ENSMUSG00000089699</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm19938</th>\n",
       "      <td>ENSMUSG00000102331</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm37381</th>\n",
       "      <td>ENSMUSG00000102343</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rp1</th>\n",
       "      <td>ENSMUSG00000025900</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC124606.1</th>\n",
       "      <td>ENSMUSG00000095523</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC133095.2</th>\n",
       "      <td>ENSMUSG00000095475</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC133095.1</th>\n",
       "      <td>ENSMUSG00000094855</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC234645.1</th>\n",
       "      <td>ENSMUSG00000095019</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC149090.1</th>\n",
       "      <td>ENSMUSG00000095041</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32285 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      gene_ids    feature_types genome\n",
       "Xkr4        ENSMUSG00000051951  Gene Expression   mm10\n",
       "Gm1992      ENSMUSG00000089699  Gene Expression   mm10\n",
       "Gm19938     ENSMUSG00000102331  Gene Expression   mm10\n",
       "Gm37381     ENSMUSG00000102343  Gene Expression   mm10\n",
       "Rp1         ENSMUSG00000025900  Gene Expression   mm10\n",
       "...                        ...              ...    ...\n",
       "AC124606.1  ENSMUSG00000095523  Gene Expression   mm10\n",
       "AC133095.2  ENSMUSG00000095475  Gene Expression   mm10\n",
       "AC133095.1  ENSMUSG00000094855  Gene Expression   mm10\n",
       "AC234645.1  ENSMUSG00000095019  Gene Expression   mm10\n",
       "AC149090.1  ENSMUSG00000095041  Gene Expression   mm10\n",
       "\n",
       "[32285 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna.var.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d342fac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = [\"rna_annot.h5ad\", \"gs.h5ad\"]\n",
    "adatas = []\n",
    "for i in files:\n",
    "    adata = sc.read_h5ad(i)\n",
    "    adatas.append(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2932ab64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[AnnData object with n_obs × n_vars = 1949 × 32285\n",
       "     obs: 'annot'\n",
       "     var: 'gene_ids', 'feature_types', 'genome'\n",
       "     uns: 'annot_colors'\n",
       "     obsm: 'spatial',\n",
       " AnnData object with n_obs × n_vars = 1949 × 55291\n",
       "     obs: 'tsse', 'n_fragment', 'frac_dup', 'frac_mito'\n",
       "     uns: 'reference_sequences'\n",
       "     obsm: 'insertion']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adatas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d91c6642",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adatas[1].obs_names = [i.split(\"-\")[0] for i in adatas[1].obs_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "77f5cbc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adatas[1] = adatas[1][adatas[0].obs_names]\n",
    "# adatas[0].obs[\"cell_type\"] = list(adatas[0].obs[\"annot\"])\n",
    "# adatas[1].obs[\"cell_type\"] = list(adatas[0].obs[\"cell_type\"])\n",
    "# adatas[1].obsm[\"spatial\"] = adatas[0].obsm[\"spatial\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ef2b27b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import numpy as np\n",
    "# for i, ad in enumerate(adatas):\n",
    "#     ad.obs_names = [j+f\"_{i}\" for j in ad.obs_names]\n",
    "#     ad.var_names_make_unique()\n",
    "# adatas[0].obs[\"batch\"] = \"RNA\"\n",
    "# adatas[1].obs[\"batch\"] = \"ATAC\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "120cc4dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# concat = sc.concat(adatas)\n",
    "# concat.obs[\"spatial1\"] =  list(concat.obsm[\"spatial\"][:,0])\n",
    "# concat.obs[\"spatial2\"] =  list(concat.obsm[\"spatial\"][:,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a1a68264",
   "metadata": {},
   "outputs": [],
   "source": [
    "# concat.write(\"concat.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2c3e553b",
   "metadata": {},
   "outputs": [],
   "source": [
    "bm = Benchmarker(R_conda_env=\"Rbase\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fd4997ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.h5ad2rds(in_file=concat, out_file=\"concat.rds\", verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7af85dce",
   "metadata": {},
   "source": [
    "Run all method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05b8d8d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.run(RDS_file_path=\"concat.rds\", H5AD_file_path=\"concat.h5ad\",\n",
    "#        save_path=\"../../../result/MultiEmbryo2/embed/\", \n",
    "#        n_cluster=7, verbose=True, workers=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7882c89",
   "metadata": {},
   "source": [
    "Evaluate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "66cc9f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\"concat.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90881750",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dict = bm.read_result(\"../../result/MultiEmbryo2/embed/\", \n",
    "                          index=list(adata.obs_names),\n",
    "                          reindex=False,\n",
    "                          save=\"../../result/MultiEmbryo2/embed_dict.pkl\",\n",
    "                          methods=bm.all_methods+[\"BindSC\", \"GLUE\", \"Monae\", \"SCALEX\", \"scConfluence\", \"SIMBA\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc7c2dad",
   "metadata": {},
   "outputs": [],
   "source": [
    "metrics = bm.cal_metrics(adata=adata, batch_key=\"batch\", label_key=\"cell_type\",\n",
    "                         res_dict=res_dict, methods=\"all\", verbose=True, rep=1,\n",
    "                         min_max_scale=False, save=\"../../result/MultiEmbryo2/metrics.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "91a7b2e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../../result/MultiEmbryo2/metrics.pkl\", \"rb\") as f:\n",
    "    metrics = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb659de3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-12-21 17:35:09 - INFO - Custom font 'Helvetica' has been set\n"
     ]
    }
   ],
   "source": [
    "bm.set_plot_params(params_dict={\"figure.dpi\": 300}, font_file_path=\"./Helvetica.ttf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72c4c76f",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = \"../../..//figures/Embryo2/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7f3296e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.plot_heatmap(metric_df=metrics[0],\n",
    "#                 save=f\"{save_dir}/summary_heatmap.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ec15d14e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from benchmarker import split_adata, transform_coord\n",
    "import numpy as np\n",
    "spatial = [ad.obsm[\"spatial\"] for ad in split_adata(adata)]\n",
    "spatial = transform_coord(spatial, vertical=True, margin_size=0.05, axis=\"y\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "098435e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# spatial = [adata.obsm[\"spatial\"]]\n",
    "# spatial = transform_coord(spatial, vertical=True,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c7314ba3",
   "metadata": {},
   "outputs": [],
   "source": [
    "bg_dict = {i:\"#D4B483\" if i in bm.spatial_methods else \"#5873a4\" for i in res_dict[\"UMAP\"].keys() }\n",
    "bg_dict[\"STADIA\"] = \"#D4B483\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cfe9956d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.plot_legend(category_lst=adata.obs[\"cell_type\"], marker=\"o\", ncol=2, save=f\"{save_dir}/annot_legend.pdf\",\n",
    "#                handletextpad=0, borderpad=0,\n",
    "#                labelspacing=0.6, columnspacing=0,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "67c98247",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.plot_legend(category_lst=adata.obs[\"cell_type\"], marker=\"o\", ncol=3, save=f\"{save_dir}/annot_legend2.pdf\",\n",
    "#                handletextpad=0, borderpad=0,\n",
    "#                labelspacing=0.6, columnspacing=0,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d992b1f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.plot_legend(category_lst=list(range(7)), marker=\"o\", ncol=3, save=f\"{save_dir}/cluster_legend.pdf\",)\n",
    "#             #    handletextpad=0, borderpad=0,\n",
    "#             #    labelspacing=0.6, columnspacing=0,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1f023b5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.plot_legend(category_lst=list(adata.obs[\"batch\"]), marker=\"o\", ncol=1,\n",
    "#             #    order=[\"Visium\", \"Stereo-seq\", \"Slide-seq V2\"],\n",
    "#                save=f\"{save_dir}/batch_legend.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "dd10427f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.plot_spatial(spatial=spatial, label_dict={\"annot\": np.array(adata.obs[\"cell_type\"]).reshape(-1,1)},\n",
    "#                 figsize=(2, 4), frameon=True, inner_gs_row=2, inner_gs_col=1, size=15, ncol=1,\n",
    "#                 xlabel=[\"Annotation\"], ylabel=None, only_show_left=True,\n",
    "#                 axis_width = 1.2, axis_color=\"lightgrey\",\n",
    "#                 outer_row_hspace=0.15, outer_col_wspace=0.1,\n",
    "#                 background_color = lambda x: bg_dict[x] if x in bg_dict else None,\n",
    "#                 xlabel_pad=0.015,\n",
    "#                 save=f\"{save_dir}/annot_spatial_plot.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "02e4c870",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import rcParams\n",
    "if len(rcParams[\"axes.prop_cycle\"].by_key()[\"color\"]) >= 10:\n",
    "        cc = rcParams[\"axes.prop_cycle\"]()\n",
    "        palette = [next(cc)[\"color\"] for _ in range(10)]\n",
    "cmap = {str(i): palette[i] for i in range(10)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "de26a8f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.plot_spatial(spatial=spatial, label_dict=res_dict[\"Cluster\"],\n",
    "#                 figsize=(18, 14), frameon=True, inner_gs_row=2, inner_gs_col=1, size=20, ncol=8,\n",
    "#                 xlabel=\"name\", ylabel=[\"RNA\", \"ATAC\"], only_show_left=True,\n",
    "#                 axis_width = 1.2, axis_color=\"lightgrey\",\n",
    "#                 # order=[\"Graspot\", \"SLAT\", \"STAligner\", \"DeepST\", \"scVI\",  \"Seurat\", \"LIGER\",],\n",
    "#                 outer_row_hspace=0.15, outer_col_wspace=0.05, inner_common_camp=False,\n",
    "#                 background_color = lambda x: bg_dict[x] if x in bg_dict else None,\n",
    "#                 xlabel_pad=0.015, palette=cmap, save_dpi=600,\n",
    "#                 save=f\"{save_dir}/all_spatial.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "31f0f18f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.plot_spatial(spatial=spatial, label_dict=res_dict[\"Cluster\"],\n",
    "#                 figsize=(10, 4.0), frameon=True, inner_gs_row=2, inner_gs_col=1, size=18, ncol=5,\n",
    "#                 xlabel=\"name\", ylabel=[\"RNA\", \"ATAC\"], only_show_left=True,\n",
    "#                 axis_width = 1.2, axis_color=\"lightgrey\",\n",
    "#                 order=[\"SLAT\", \"scVI\", \"GLUE\", \"Monae\", \"BindSC\"],\n",
    "#                 inner_row_hspace=0.18, outer_col_wspace=0.05, inner_common_camp=True,\n",
    "#                 background_color = lambda x: bg_dict[x] if x in bg_dict else None,\n",
    "#                 xlabel_pad=0.012, palette=cmap, save_dpi=600,\n",
    "#                 save=f\"{save_dir}/unpaired_spatial.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e0904559",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.plot_umap(embed_dict=res_dict[\"UMAP\"], batch_dict=res_dict[\"Batch\"], annot_list=list(adata.obs[\"cell_type\"]), \n",
    "#              figsize=(14, 12), frameon=True, inner_gs_row=2, inner_gs_col=1, size=3, ncol=8,\n",
    "#              ylabel=[\"Batch\", \"Cell type\"], only_show_top=False,  xlabel=\"name\", only_show_left=True,\n",
    "#              background_color = lambda x: bg_dict[x] if x in bg_dict else None,\n",
    "#             #  order=[\"Unintegrated\", \"SLAT\", \"Seurat\", \"GLUE\", \"SCALEX\", \"BindSC\"],\n",
    "#              axis_width=1.2, axis_color=\"lightgrey\",\n",
    "#              outer_col_wspace=0.05, save_dpi=600, outer_row_hspace=0.15,\n",
    "#              ylabel_pad=0.018, xlabel_pad=0.015, merge=True, merge_margin_size=0.4,\n",
    "#              save=f\"{save_dir}/all_umap.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "1f9c5bb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "paired_methods = [\"SpatialGlue\", \"COSMOS\", \"MISO\", \"PRESENT\"]\n",
    "embed, cluster = {}, {}\n",
    "for i in paired_methods:\n",
    "    embed[i] = pd.read_csv(f\"../../result/MultiEmbryo2/embed/embed_{i}.csv\", index_col=0)\n",
    "    cluster[i] = pd.read_csv(f\"../../result/MultiEmbryo2/embed/cluster_{i}.csv\", index_col=0)\n",
    "    if i == \"PRESENT\":\n",
    "        # pass\n",
    "        cluster[i].index = [i.split(\"-\")[0] for i in cluster[i].index]\n",
    "        embed[i].index = cluster[i].index.copy()\n",
    "        embed[i] = embed[i].reindex(rna.obs_names)\n",
    "        cluster[i] = cluster[i].reindex(rna.obs_names)\n",
    "    else:\n",
    "        embed[i].index = rna.obs_names.copy()\n",
    "    cluster[i] = np.array(cluster[i].iloc[:,0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "384afbb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "spatial = [split_adata(adata)[0].obsm[\"spatial\"]]\n",
    "spatial = transform_coord(spatial, vertical=True, margin_size=0.05, axis=\"y\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "c3fcdb8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.plot_spatial(spatial=spatial,\n",
    "#                 label_dict=cluster,\n",
    "#                 figsize=(4.3, 4.5), frameon=True, inner_gs_row=1, inner_gs_col=1, size=18, ncol=2,\n",
    "#                 xlabel=\"name\",\n",
    "#                 only_show_left=True,\n",
    "#                 axis_width = 1.2, axis_color=\"lightgrey\",\n",
    "#                 order=[\"SpatialGlue\", \"MISO\", \"COSMOS\", \"PRESENT\"],\n",
    "#                 outer_row_hspace=0.2, outer_col_wspace=0.05, inner_common_camp=True,\n",
    "#                 background_color = lambda x: \"#D4B483\",\n",
    "#                 xlabel_pad=0.012,\n",
    "#                 palette=cmap,\n",
    "#                 save_dpi=600,\n",
    "#                 save=f\"{save_dir}/paired_spatial.pdf\"\n",
    "#                 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "a7ad4f34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.plot_umap(embed_dict=res_dict[\"UMAP\"], batch_dict=res_dict[\"Batch\"], annot_list=list(adata.obs[\"cell_type\"]), \n",
    "#              figsize=(10.5, 3.9), frameon=True, inner_gs_row=2, inner_gs_col=1, size=3, ncol=6,\n",
    "#              ylabel=[\"Batch\", \"Cell type\"], only_show_top=False,  xlabel=\"name\", only_show_left=True,\n",
    "#              background_color = lambda x: bg_dict[x] if x in bg_dict else None,\n",
    "#              order=[\"Unintegrated\", \"SLAT\", \"scVI\", \"GLUE\", \"Monae\", \"BindSC\"],\n",
    "#              axis_width=1.2, axis_color=\"lightgrey\",\n",
    "#              outer_col_wspace=0.05, save_dpi=600,\n",
    "#              ylabel_pad=0.018, xlabel_pad=0.015, merge=True, merge_margin_size=0.4,\n",
    "#              save=f\"{save_dir}/unpaired_umap.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "107c1508",
   "metadata": {},
   "outputs": [],
   "source": [
    "# umap = {}\n",
    "# for i in embed.keys():\n",
    "#     t_adata = sc.AnnData(X=embed[i])\n",
    "#     sc.pp.neighbors(t_adata, use_rep=\"X\")\n",
    "#     sc.tl.umap(t_adata)\n",
    "#     umap[i] = t_adata.obsm[\"X_umap\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "a20f6160",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bg_dict = {i:\"#D4B483\" if i not in bm.spatial_methods else \"#5873a4\" for i in embed.keys() }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "63350201",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.plot_umap(embed_dict=umap, annot_list=list(rna.obs[\"annot\"]), \n",
    "#              figsize=(3.8, 4.1), frameon=True, inner_gs_row=1, inner_gs_col=1, size=5, ncol=2,\n",
    "#              ylabel=[\"Cell type\"], only_show_top=False,  xlabel=\"name\", only_show_left=True,\n",
    "#              background_color = lambda x: bg_dict[x] if x in bg_dict else None,\n",
    "#             order=[\"SpatialGlue\", \"MISO\", \"COSMOS\", \"PRESENT\"],\n",
    "#              axis_width=1.2, axis_color=\"lightgrey\",\n",
    "#              outer_col_wspace=0.05, save_dpi=600,  outer_row_hspace=0.2,\n",
    "#              ylabel_pad=0.018, xlabel_pad=0.012, merge=False, merge_margin_size=0.4,\n",
    "#              save=f\"{save_dir}/paired_umap.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6473087b",
   "metadata": {},
   "outputs": [],
   "source": [
    "batch_metric = metrics[0].copy()\n",
    "batch_metric = batch_metric.loc[:,[\"Silhouette batch\",\"iLISI\", \"KBET\", \"Graph connectivity\", \"PCR comparison\", \"Batch correction\"]]\n",
    "batch_metric.loc[\"Metric Type\"] = [i.replace(\" \", \"\\n\") for i in batch_metric.loc[\"Metric Type\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b21dbb76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.plot_heatmap(metric_df=batch_metric, save=f\"{save_dir}/summary_batch_heatmap_all.pdf\", total_name=\"Batch correction\", )#insert_marker_row=8, show_top=7, show_bottom=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e2eae2b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "paired_res_dict = {}\n",
    "paired_res_dict[\"Cluster\"] = cluster.copy()\n",
    "paired_res_dict[\"Embed\"] = embed.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "699f23a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna.obs[\"batch\"] = [\"b1\" if i<1200  else \"b2\" for i in range(rna.shape[0])] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "850f6270",
   "metadata": {},
   "outputs": [],
   "source": [
    "bio_metric = metrics[0].copy()\n",
    "bio_metric = bio_metric.loc[:,['CHAOS', 'PAS', 'Isolated labels', 'NMI', 'ARI', 'Silhouette label', 'cLISI', 'Domain continuity', 'Bio conservation']]\n",
    "# bio_metric.loc[\"Metric Type\"] = [i.replace(\" \", \"\\n\") for i in bio_metric.loc[\"Metric Type\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c1446e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "paired_metric = bm.cal_metrics(adata=rna, batch_key=\"batch\", label_key=\"annot\",\n",
    "               res_dict=paired_res_dict, methods=\"all\", verbose=True, rep=1,\n",
    "               min_max_scale=False, save=\"../../result/MultiEmbryo2/paired_metrics.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e65f4fbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "bio_paired_metirc = paired_metric[0].loc[:,['CHAOS', 'PAS', 'Isolated labels', 'NMI', 'ARI', 'Silhouette label',\n",
    "       'cLISI', 'Domain continuity', 'Bio conservation']]\n",
    "bio_paired_metirc = bio_paired_metirc.drop(\"Metric Type\")\n",
    "bio_metric_all = pd.concat((bio_metric, bio_paired_metirc))#.sort_values(\"Bio conservation\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d9d1f176",
   "metadata": {},
   "outputs": [],
   "source": [
    "bio_metric_all.index = [\"*\"+i if i in paired_methods else i for i in bio_metric_all.index]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "1468b9f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bm.plot_heatmap(metric_df=bio_metric_all, save=f\"{save_dir}/summary_bio_heatmap.pdf\", total_name=\"Bio conservation\", insert_marker_row=8, show_top=7, show_bottom=5)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "benchmark_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
